suite/suite.go

```diff
@@ -96,19 +96,19 @@ func failOnPanic(t *testing.T, r interface{}) {
 func (suite *Suite) Run(name string, subtest func()) bool {
 	oldT := suite.T()
 
-	if setupSubTest, ok := suite.s.(SetupSubTest); ok {
-		setupSubTest.SetupSubTest()
-	}
+	return oldT.Run(name, func(t *testing.T) {
+		suite.SetT(t)
```

Comment:
Should we add `defer receiveAndFailOnPanic(t)` here to ensure that `TearDownSubTest` and `TearDownSuite` will be executed even if a panic occurs in a sub-test?

Reply:
I assume you mean `recoverAndFailOnPanic(t)`.
Since the `TearDownSubTest` is deferred, it will be called even if a panic occurs in a subtest.
The same for the `TearDownSuite`: It is deferred further up in the call stack.

But your suggestion might still provide advantages: If a `SetupSubTest` or a `TearDownSubTest` panics, the test will fail "on the subtest it belongs to" and not on a parent test. Also the next subtests will still be able to execute (if that is the wanted behavior).
I would just add it for now and if there are any concerns please let me know.

Reply:
> The same for the TearDownSuite: It is deferred further up in the call stack.

To be clear, I tested against your branch and if a panic occurs in a subtest, the `TearDownSuite` will not be executed. (no `recoverAndFailOnPanic(t)`), my test code: https://go.dev/play/p/UaGXEUZgEU2


> Since the TearDownSubTest is deferred, it will be called even if a panic occurs in a subtest.

As you said below, if a panic occurs in a subtest, it will also terminate the entire subtest process. I think executing the next subtest should be expected behavior because the suite-level tests have that behavior and there is no reason for the subtest to be inconsistent with it.

In fact, there is [a PR](https://github.com/stretchr/testify/pull/1293) trying to solve this problem. Since you made changes here, it would be great if you could fix it by the way. This PR has a test case, would you mind moving it over for completeness? It may need to be modified to be more complete.

Reply:
Oh man! You're right! Thanks for your testing code, that helped me reproducing it. You might already know, but I found out that this probably has to do with subtests being run in goroutines which I did not know (see: https://github.com/golang/go/issues/20394).

I agree with the consistency of the panic-terminate-behavior for tests and subtests.

I moved your test code into this PR. Thanks for your help. Would be great if you have a look to check if I missed something.

Reply:
LGTM, well done.

Reply:
@linusbarth @xwjdsh Well done, guys! Great collaboration!